Dockerfile

```diff
@@ -0,0 +1,39 @@
+# syntax=docker/dockerfile:1
+
+# Stage 1 - Build the toolchain
+FROM ubuntu:18.04
+ARG N64_INST=/n64_toolchain
+ENV N64_INST=${N64_INST}
+
+# install dependencies
+RUN apt-get update
+RUN apt-get install -yq wget bzip2 gcc g++ make file libmpfr-dev libmpc-dev zlib1g-dev texinfo git gcc-multilib
+
+# Build
+COPY ./tools/build-toolchain.sh /tmp/tools/build-toolchain.sh
+WORKDIR /tmp/tools
+RUN ./build-toolchain.sh
+
+# Strip executables
+RUN find ${N64_INST}/bin -type f | xargs strip
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/plugin/gengtype
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/liblto_plugin.so.0.0.0
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/liblto_plugin.so.0
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/lto-wrapper
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/collect2
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/cc1plus
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/cc1
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/install-tools/fixincl
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/lto1
+RUN rm -rf ${N64_INST}/share/locale/*
+
+# Stage 2 - Prepare minimal image
+FROM ubuntu:18.04
+ARG N64_INST=/n64_toolchain
+ENV N64_INST=${N64_INST}
```

Comment:
TIL multi stage builds!

How big is the final image now?

Reply:
Not very different, a little smaller: 619MB
But the practical implications of this is much important as we don't have to run the whole build if we write a failing strip command now. The previous steps will be cached.

Reply:
I was previously debugging the image build on my pc by first converting a big `... && ...` into docker commands to enable caching and back to a huge bash command after I'm done so that final image is small.

There was even someone e-mailing me that said he converted every step of the toolchain build into docker `RUN`s to be able to tweak it :)